stdlib/iter.mli

```diff
@@ -0,0 +1,42 @@
+(**************************************************************************)
+(*                                                                        *)
+(*                                 OCaml                                  *)
+(*                                                                        *)
+(*                 Simon Cruanes                                          *)
+(*                                                                        *)
+(*   Copyright 2001 Institut National de Recherche en Informatique et     *)
+(*     en Automatique.                                                    *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+(* Module [Iter]: functional iterators *)
+
+(** @since NEXT_RELEASE *)
+
+type ('a, 's) step =
+  | Done
+  | Skip of 's
+  | Yield of 'a * 's
+
+type +_ t =
+  | Sequence : 's * ('s -> ('a,'s) step) -> 'a t
+
+val empty : 'a t
+
+val return : 'a -> 'a t
+
+val map : ('a -> 'b) -> 'a t -> 'b t
+
+val filter : ('a -> bool) -> 'a t -> 'a t
+
+val filter_map : ('a -> 'b option) -> 'a t -> 'b t
+
+val flat_map : ('a -> 'b t) -> 'a t -> 'b t
+
+val fold_left : ('a -> 'b -> 'a) -> 'a -> 'b t -> 'a
```

Comment:
The reason for preferring that signature is that it composes well.  In particular you can directly express nested folds like `List.fold (Int_set.fold f)`, which also works across containers.  In this case, e.g. if you have two sequences you wish to collect as a set, where the latter contains lists of elements:
```ocaml
Int_set.empty
  |> Seq.fold Int_set.add seq1
  |> Seq.fold (List.fold Int_set.add) seq2
```